plonky2/src/plonk/config.rs

```diff
@@ -19,62 +19,114 @@ use crate::field::goldilocks_field::GoldilocksField;
 use crate::hash::hash_types::{HashOut, RichField};
 use crate::hash::hashing::PlonkyPermutation;
 use crate::hash::keccak::KeccakHash;
-use crate::hash::poseidon2_bn254::Poseidon2BN254;
 use crate::hash::poseidon::PoseidonHash;
 use crate::iop::target::{BoolTarget, Target};
 use crate::plonk::circuit_builder::CircuitBuilder;
+use ark_bn254::Fr as BN254Fr;
+use ark_ff::{One, Zero};
+use crate::hash::poseidon2_bn254::{bytes_to_felts, felts_to_bytes, Poseidon2BN254};
 
 pub trait GenericHashOut<F: RichField>:
-    Copy + Clone + Debug + Eq + PartialEq + Send + Sync + Serialize + DeserializeOwned
+Copy + Clone + Debug + Eq + PartialEq + Send + Sync + Serialize + DeserializeOwned
 {
     fn to_bytes(&self) -> Vec<u8>;
     fn from_bytes(bytes: &[u8]) -> Self;
 
-    fn to_vec(&self) -> Vec<F>;
+    fn to_vec(&self) -> Vec<GenericField<F>>;
+}
+
+/// generic field enum - supports only 2 fields for now
+/// Supported fields: Goldilocks , BN254 Fr (from Arkworks)
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum GenericField<F: RichField> {
+    Goldilocks(F),
+    BN254(BN254Fr),
+}
+
+/// hasher field trait to cover fields in `GenericField` enum
+pub trait HasherField: Default + Sized + Copy + Debug + Eq + PartialEq + Sync + Send {
+    fn get_one() -> Self;
+    fn get_zero() -> Self;
+    fn to_bytes(&self) -> Vec<u8>;
+
+    fn from_bytes(b: &[u8]) -> Self;
+
+}
+
+/// BN254 as Hasherfield
+impl HasherField for BN254Fr {
+    fn get_one() -> Self {
+        BN254Fr::one()
+    }
+
+    fn get_zero() -> Self {
+        BN254Fr::zero()
+    }
+
+    fn to_bytes(&self) -> Vec<u8> {
```

Comment:
I would add some tests that `to_bytes` and `from_bytes` are the inverse of each other and behave as expected

btw as in cryptography, we usually have little-endian and big-endian mixed over all the place, I found it useful to postfix all such conversation function by `_LE` or `_BE` so that's it's always clear which one we deal with (in rust maybe it's more common to use lowercase `_le` and `_be`. I don't have a strong opinion that, though upper case is easier to figure out to me when reading the code)